File: packages/js-dpp/test/unit/document/stateTransition/data/validatePartialCompoundIndices.spec.js
const validatePartialCompoundIndices = require('../../../../../lib/document/stateTransition/validation/data/validatePartialCompoundIndices');
const InconsistentCompoundIndexDataError = require('../../../../../lib/errors/InconsistentCompoundIndexDataError');

const getDocumentsFixture = require('../../../../../lib/test/fixtures/getDocumentsFixture');
const getContractFixture = require('../../../../../lib/test/fixtures/getDataContractFixture');
const getDocumentTransitionsFixture = require('../../../../../lib/test/fixtures/getDocumentTransitionsFixture');

const ValidationResult = require('../../../../../lib/validation/ValidationResult');
const { expectValidationError } = require('../../../../../lib/test/expect/expectError');

describe('validatePartialCompoundIndices', () => {
  let documents;
  let documentTransitions;
  let dataContract;
  let ownerId;

  beforeEach(() => {
    dataContract = getContractFixture();
    ownerId = '5bGvpuiXVW3yK8np1u51Y2LFk2WCvztpa8yYy6VJpguc';
  });

  it('should return invalid result if compound index contains not all fields', () => {
    const document = getDocumentsFixture(dataContract)[9];
    document.set('lastName', undefined);

    documents = [document];
    documentTransitions = getDocumentTransitionsFixture({
      create: documents,
    });

    const result = validatePartialCompoundIndices(ownerId, documentTransitions, dataContract);

    expectValidationError(result, InconsistentCompoundIndexDataError);

    const { optionalUniqueIndexedDocument } = dataContract.getDocuments();
    const [error] = result.getErrors();

    expect(error.getIndexDefinition()).to.deep.equal(
      optionalUniqueIndexedDocument.indices[1],
    );
    expect(error.getDocumentType()).to.equal('optionalUniqueIndexedDocument');
  });

  it('should return valid result if compound index contains no fields', () => {
    const document = getDocumentsFixture(dataContract)[8];
    document.setData({ });

    documents = [document];

    documentTransitions = getDocumentTransitionsFixture({
      create: documents,
    });

    const result = validatePartialCompoundIndices(ownerId, documentTransitions, dataContract);

    expect(result).to.be.an.instanceOf(ValidationResult);
    expect(result.isValid()).to.be.true();
  });

  it('should return valid result if compound index contains all fields', () => {
    documents = [getDocumentsFixture(dataContract)[8]];
    documentTransitions = getDocumentTransitionsFixture({
      create: documents,
    });

    const result = validatePartialCompoundIndices(ownerId, documentTransitions, dataContract);

    expect(result).to.be.an.instanceOf(ValidationResult);
    expect(result.isValid()).to.be.true();
  });
});

File: packages/rs-dpp/src/tests/document/state_transitions/validation/validate_partial_compound_indices_spec.rs
use platform_value::Value;

use crate::{
    consensus::{basic::BasicError, ConsensusError},
    data_contract::DataContract,
    document::{
        document_transition::{Action, DocumentTransitionObjectLike},
        state_transition::documents_batch_transition::validation::basic::validate_partial_compound_indices::*,
    },
    tests::fixtures::{
        get_data_contract_fixture, get_document_transitions_fixture,
    },
    validation::ValidationResult,
};
use crate::document::ExtendedDocument;
use crate::tests::fixtures::get_extended_documents_fixture;

struct TestData {
    data_contract: DataContract,
    documents: Vec<ExtendedDocument>,
}

fn setup_test() -> TestData {
    let data_contract = get_data_contract_fixture(None);
    let documents =
        get_extended_documents_fixture(data_contract.clone()).expect("documents should be created");

    TestData {
        data_contract,
        documents,
    }
}

#[test]
fn should_return_invalid_result_if_compound_index_contains_not_all_fields() {
    let TestData {
        data_contract,
        mut documents,
    } = setup_test();
    let mut document = documents.remove(9);
    document
        .properties_as_mut()
        .remove("lastName")
        .expect("lastName property should exist and be removed");

    let documents_for_transition = vec![document];
    let raw_document_transitions: Vec<Value> =
        get_document_transitions_fixture([(Action::Create, documents_for_transition)])
            .into_iter()
            .map(|dt| {
                dt.to_object()
                    .expect("the transition should be converted to object")
            })
            .collect();
    let result = validate_partial_compound_indices(raw_document_transitions.iter(), &data_contract)
        .expect("should return validation result");

    let basic_error = get_basic_error(&result, 0);

    assert!(!result.is_valid());
    assert_eq!(1021, result.consensus_errors[0].code());
    match basic_error {
        BasicError::InconsistentCompoundIndexDataError(err) => {
            assert_eq!(
                err.document_type(),
                "optionalUniqueIndexedDocument".to_string()
            );
            assert_eq!(
                err.index_properties(),
                vec![
                    "$ownerId".to_string(),
                    "firstName".to_string(),
                    "lastName".to_string()
                ]
            );
        }
        _ => panic!(
            "Expected InconsistentCompoundIndexDataError, got {}",
            basic_error
        ),
    }
}

#[test]
fn should_return_valid_result_if_compound_index_contains_nof_fields() {
    let TestData {
        data_contract,
        mut documents,
    } = setup_test();
    let mut document = documents.remove(8);
    document.properties_as_mut().clear();

    let documents_for_transition = vec![document];
    let raw_document_transitions: Vec<Value> =
        get_document_transitions_fixture([(Action::Create, documents_for_transition)])
            .into_iter()
            .map(|dt| {
                dt.to_object()
                    .expect("the transition should be converted to object")
            })
            .collect();
    let result = validate_partial_compound_indices(raw_document_transitions.iter(), &data_contract)
        .expect("should return validation result");
    assert!(result.is_valid());
}

#[test]
fn should_return_valid_result_if_compound_index_contains_all_fields() {
    let TestData {
        data_contract,
        mut documents,
    } = setup_test();
    let document = documents.remove(8);
    let documents_for_transition = vec![document];
    let raw_document_transitions: Vec<Value> =
        get_document_transitions_fixture([(Action::Create, documents_for_transition)])
            .into_iter()
            .map(|dt| {
                dt.to_object()
                    .expect("the transition should be converted to object")
            })
            .collect();
    let result = validate_partial_compound_indices(raw_document_transitions.iter(), &data_contract)
        .expect("should return validation result");
    assert!(result.is_valid());
}

fn get_basic_error(result: &ValidationResult<()>, error_number: usize) -> &BasicError {
    match result
        .consensus_errors
        .get(error_number)
        .expect("error should be found")
    {
        ConsensusError::BasicError(basic_error) => basic_error,
        _ => panic!(
            "error '{:?}' isn't a basic error",
            result.consensus_errors[error_number]
        ),
    }
}
